File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* findmin(TreeNode* root){
    while(root->left)root = root->left;
    return root;
}


    TreeNode* deleteNode(TreeNode* root, int val) {
        if(!root)return root;
    //searching
    if(val < root->val){
        root->left = deleteNode(root->left , val);
        
    }else if(val  > root->val){
        root->right = deleteNode(root->right , val);
    }else{
        //mil gaya 
        //check for their child
        //only one child
        if(!root->left){
            TreeNode* temp = root->right;
            delete root;
            return temp;
        }else if(!root->right){
            TreeNode* temp = root->left ;
            delete root ;
            return temp;
        }else{
            //have 2 child
            TreeNode* temp = findmin(root->right);
            root->val = temp->val;
            root->right = deleteNode(root->right ,temp->val );
            
        }
    }
    return root;
    }
};

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        if (root == null) return null;

        if (key < root.val) {
            // Target is in the left subtree
            root.left = deleteNode(root.left, key);
        } else if (key > root.val) {
            // Target is in the right subtree
            root.right = deleteNode(root.right, key);
        } else {
            // Node to delete found!
            
            // Case 1 & 2: Leaf or one child
            if (root.left == null) return root.right;
            if (root.right == null) return root.left;

            // Case 3: Two children
            // Find the inorder successor (smallest in the right subtree)
            TreeNode successor = findMin(root.right);
            root.val = successor.val;
            // Delete the inorder successor
            root.right = deleteNode(root.right, successor.val);
        }
        return root;
    }

    private TreeNode findMin(TreeNode node) {
        while (node.left != null) {
            node = node.left;
        }
        return node;
    }
}